src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2043,11 +2043,11 @@ protected override void AfterLeftChildHasBeenVisited(BoundBinaryOperator binary)
 
             // learn from non-null constant
             BoundExpression operandComparedToNonNull = null;
-            if (isNonNullConstant(binary.Left))
+            if (leftType.IsNotNull && rightType.MayBeNull && binary.Right.ConstantValue?.IsNull != true)
```

Comment:
The point here is to avoid attempting to learn from a comparison between a null literal and a not-null expression so that we can continue on to the part where we learn from comparison to null.

Reply:
Please add some test where the expression is non-null due to a conversion.
```
class A
{
  // can implicitly convert A? to B
}
class B { }
...
if ( /*implicit conversion here*/ maybeNullA == nonNullB) maybeNullA.ToString(); // still expecting warning
```

---
In reply to: [287527897](https://github.com/dotnet/roslyn/pull/35953#discussion_r287527897) [](ancestors = 287527897)

Reply:
It looks like a comparison between values of type A and B will not cause a user-defined implicit conversion to occur--you have to cast. Will add as similar of a test to what you've described as possible.

https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAmARgFgAoUnAZgAJ8qBBKgb1KtZupwIDYq0wAHDGgDGaGFQD2/RBBgSEVAEIAKBhACUVALwA+KgDs4AdyXL1AblIBfUuQ54lTa7ZKUaDgMJOSbdjQAsVACyZt6+vmraBgCuGBiWPuGsilTAUYYmKhYsSbwAZlTKENpaqZqMVDlJNolJVeFoBcrKiurFWqXA5fW+Nb3WQA===